### ISSUE TITLE
Implement a code generator that takes Smithy spec and generates zio.http.api.EndpointSpec values

### ISSUE THREAD

**Issue by jdegoes** [zio/zio-http#1522]
**Is your feature request related to a problem? Please describe.**

Some companies begin with Smithy specifications, and then proceed to implement a server that satisfies a given Smithy specification; then they will modify the Smithy specification when a change is necessary, and then re-implement the server.

**Describe the solution you'd like**

In order to support this _spec-first_ driven approach to API development (rather than _code-first_, which is the paradigm already implemented within the `zio.http.api` package), we will need a code generator, which can read in a Smithy specification, and generate the corresponding `zio.http.api.EndpointSpec ` values that declaratively describe the APIs.

Then a user can implement handlers based on the code-generated `EndpointSpec ` values.

**Comment by jdegoes** [comment]
Possibly, this could be done as a plug-in to `smithy4s`.

**Comment by yisraelU** [comment]
I can take on creating a PR in this repo, to have a zio-http interpreter(Smithy4s terminology).
I would wait until Smithy4s 0.18  version is released as it will make it a lot easier.
Happy to discuss further

**Comment by saadhashmi91** [comment]
Hi, I was looking into option of using a recursion scheme(s)/fix-point  parser for Smithy IDL using its ABNF spec. In particular [skeuomorph](https://github.com/higherkindness/skeuomorph)  has an implementation of transforming protobuf schemas. If a similar approach can be used for the parser and also leveraging smithy4s for the spec. implementation.

**Comment by adamgfraser** [comment]
@saadhashmi91 We would like to do this without any third party dependencies.

**Comment by sideeffffect** [comment]
@adamgfraser just to be clear, do you allow other ZIO Organization dependencies, like zio-schema?

**Comment by adamgfraser** [comment]
@sideeffffect If that functionality is necessary for something we are trying to do. For example ZIO Schema is already a dependency since it is necessary to describe arbitrary types as values as part of the endpoints API and we would basically have to implement it ourselves otherwise. But in general would not want to bring in an additional dependency if it was merely a "nice to have".

**Comment by yisraelU** [comment]
> Hi, I was looking into option of using a recursion scheme(s)/fix-point  parser for Smithy IDL using its ABNF spec. In particular [skeuomorph](https://github.com/higherkindness/skeuomorph)  has an implementation of transforming protobuf schemas. If a similar approach can be used for the parser and also leveraging smithy4s for the spec. implementation.

What would be the point of implementing a parser if you are trying to leverage smithy4s codegen?
Also Smithy already has a parser and AST to represent it. I would recommend using that

**Comment by saadhashmi91** [comment]
@yisraelU Cool, I didn't look into it in detail.

**Comment by googley42** [comment]
@yisraelU I know you did some work on a smithy interpreter for zio-http - but were waiting for a new smithy 0.18 release for some extra functionality - did the release happen? were you able to use it? If so then we may be in a position to start evaluating soon

**Comment by yisraelU** [comment]
@googley42 yes.
It was released not too long ago. Unfortunately it is taking a bit longer as I need to refactor for the rc3 version of ziohttp which has quite a few breaking changes

**Comment by yisraelU** [comment]
I just wanted to drop a note that I have published https://yisraelu.github.io/smithy4s-zio/

**Comment by promisingcoder** [comment]
/attempt #1522

<div id="algora-attempt" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clgm2uqyh0000l70fp01n2m29/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by 987Nabil** [comment]
@promisingcoder I will implement this ticket or at least find a strategy to do so together with @jdegoes on [stream](https://www.meetup.com/zio-meetup-berlin/events/302081281/) on the 26.07.

In general, this is not an easy task and I think you wrote in discord that you are a Scala beginner. Maybe you should try to pick some smaller issues first to get some experience.

**Comment by promisingcoder** [comment]
@987Nabil I currently made a parser it is still in the beginning that takes in the smithy file and generates specs for a zio server 
and the server is running !! 
it's simple and in the beginning 
I intend to implement this feature 
I understand your concern but I intend to carry on

**Comment by 987Nabil** [comment]
/attempt 1522

<div id="algora-attempt" />

| [Algora profile](https://console.algora.io/@/987Nabil) | Completed bounties | Tech | Active attempts | Options |
| --- | --- | --- | --- | --- |
| @987Nabil | &nbsp;&nbsp;&nbsp;56 ZIO bounties <br/> + 1 bounty from 1 project | <div align="center">Scala</div> | <div align="center"></div> | [Cancel attempt](https://console.algora.io/api/bounties/clgm2uqyh0000l70fp01n2m29/cancel-attempt) |

**Comment by jivanic-demystdata** [comment]
@987Nabil Is this schedules to be part of `v3.0.0`? 
I want to introduce Smithy at work and we're based on zio-http. That'd be amazing to have this support for `v3..0.0` ❤️

**Comment by 987Nabil** [comment]
@jivanic-demystdata yes. I'll be on the zymposium on Friday and take a look with John on how to translate the formal syntax to code.

**Comment by 987Nabil** [comment]
@YanisOUALAN I already started and am half way through. That's why this task is assigned to me

**Comment by soujiro32167** [comment]
Reminder: [this project](https://yisraelu.github.io/smithy4s-zio/) already produces servers and clients in zio-http from Smithy spec

**Comment by PeterWanjihia** [comment]
@jdegoes  is this issue still available  I think I can work on it  on assignement

**Comment by PeterWanjihia** [comment]
@jdegoes  would appreciate a clarification from your end before I start since its not a recent issue

**Comment by JuanTorchia** [comment]
Hi! Is this bounty still active? I'm interested in working on it. 
Can someone confirm if this is still needed?

**Comment by nicefacer** [comment]
/attempt #1522

Hey @jdegoes and @987Nabil , can you give update and reply to these questions, i want to tackle this task:

Smithy feature scope: do we need event streams/WebSockets or just unary HTTP?

Target zio-http version/Scala version to lock API mapping?

Preferred auth coverage (basic/bearer/apiKey/custom identity resolvers)?

Should we also emit client facades/tests, or server-only?

How to map Smithy errors to HTTP statuses—follow smithy4s defaults or allow config?

Performance constraints on codegen (acceptable model size and build time budget)?

**Comment by shoaibansari5398** [comment]
/attempt #3893 

/claim #3893 

PR Link - https://github.com/zio/zio-http/pull/3893
